dbms/src/IO/Compression/CompressionCodecDelta.cpp

```diff
@@ -0,0 +1,332 @@
+#include <IO/Compression/CompressionInfo.h>
+#include <common/likely.h>
+#include <common/unaligned.h>
+
+#if defined(__x86_64__) && defined(__AVX2__)
+#include <immintrin.h>
+#endif
+
+#ifdef TIFLASH_ENABLE_AVX_SUPPORT
+#include <common/mem_utils_opt.h>
+ASSERT_USE_AVX2_COMPILE_FLAG
+#endif
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int CANNOT_COMPRESS;
+extern const int CANNOT_DECOMPRESS;
+} // namespace ErrorCodes
+
+CompressionCodecDelta::CompressionCodecDelta(UInt8 delta_bytes_size_)
+    : delta_bytes_size(delta_bytes_size_)
+{}
+
+UInt8 CompressionCodecDelta::getMethodByte() const
+{
+    return static_cast<UInt8>(CompressionMethodByte::Delta);
+}
+
+namespace
+{
+
+template <typename T>
+void compressDataForType(const char * source, UInt32 source_size, char * dest)
+{
+    if (source_size % sizeof(T) != 0)
+        throw Exception(
+            ErrorCodes::CANNOT_COMPRESS,
+            "Cannot compress with Delta codec, data size {} is not aligned to {}",
+            source_size,
+            sizeof(T));
+
+    T prev_src = 0;
+    const char * const source_end = source + source_size;
+    while (source < source_end)
+    {
+        T curr_src = unalignedLoad<T>(source);
+        unalignedStore<T>(dest, curr_src - prev_src);
+        prev_src = curr_src;
+
+        source += sizeof(T);
+        dest += sizeof(T);
+    }
+}
+
+#if defined(__x86_64__) && defined(__AVX2__)
+void compressDataFor32bits(const UInt32 * __restrict__ source, UInt32 source_size, UInt32 * __restrict__ dest)
+{
+    __m256i prev = _mm256_setzero_si256();
+    size_t i = 0;
+    for (; i < source_size / 8; ++i)
+    {
+        // curr = {a0, a1, a2, a3, a4, a5, a6, a7}
+        auto curr = _mm256_loadu_si256(reinterpret_cast<const __m256i *>(source) + i);
+        // x0 = {prev[7], a0, a1, a2, a3, a4, a5, a6}
+        const __m256i p_curr = {0x0000000000000007, 0x0000000200000001, 0x0000000400000003, 0x0000000600000005};
+        const __m256i p_prev = {0x0000000700000007, 0x0000000700000007, 0x0000000700000007, 0x0000000700000007};
+        auto x0 = _mm256_blend_epi32(
+            _mm256_permutevar8x32_epi32(curr, p_curr), // {a7, a0, a1, a2, a3, a4, a5, a6}
+            _mm256_permutevar8x32_epi32(prev, p_prev), // {prev[7], prev[7], ...}
+            0b00000001);
+        auto delta = _mm256_sub_epi32(curr, x0);
+        _mm256_storeu_si256(reinterpret_cast<__m256i *>(dest) + i, delta);
+        prev = curr;
+    }
+    UInt32 lastprev = _mm256_extract_epi32(prev, 7);
+    for (i = 8 * i; i < source_size; ++i)
+    {
+        UInt32 curr = source[i];
+        dest[i] = curr - lastprev;
+        lastprev = curr;
+    }
+}
+
+void decompressDataFor32bits(const UInt32 * __restrict__ source, UInt32 source_size, UInt32 * __restrict__ dest)
+{
+    __m128i prev = _mm_setzero_si128();
+    size_t i = 0;
+    for (; i < source_size / 4; i++)
+    {
+        auto curr = _mm_lddqu_si128(reinterpret_cast<const __m128i *>(source) + i);
```

Comment:
Why use 128bits rather than 256 bits here?

Reply:
Using 256 bits is too complex, may not be more effective than 128 bits.